lib/cluster.js

```diff
@@ -658,15 +663,21 @@ function workerInit() {
     if (accepted) server.onconnection(0, handle);
   }
 
-  Worker.prototype.disconnect = function() {
+  Worker.prototype.disconnect = function(fromParent) {
     this.suicide = true;
     let waitingCount = 1;
 
     function checkWaitingCount() {
       waitingCount--;
       if (waitingCount === 0) {
-        send({ act: 'suicide' });
-        process.disconnect();
+        // if called from worker, wait for ack to be sure suicide property is
```

Comment:
How about:

`// If disconnect is worker initiated, wait for ack to be sure suicide is properly set in the master`